### PR TITLE
docs: added examples for setup and teardown apis

### PR DESCRIPTION
**Pull request by dammy001**

### CODE REVIEW

**Comment by patak-dev** [comment]
Maybe we could add a one line comment in each snippet explaining what is achieved? Like "count the number of tests", and "done with the first test" (maybe we could look for a better example for afterEach 🤔)

**Comment by dammy001** [comment]
@patak-dev okay, I'll add that now

**Comment by dammy001** [comment]
@patak-dev check it out and also suggest

File: docs/api/index.md [hunk 1]
```diff
@@ -166,24 +166,74 @@ These global functions allows you to hook into the life cycle of tests to avoid
 
 Register a callback to be called before each test in the current context.
 
-**Type:** `beforeEach(fn: () => Awaitable<void>, timeout: number)`
+**Type:** `beforeEach(fn: () => Awaitable<void>, timeout?: number)`
+
+```ts
+import { beforeEach } from 'vitest'
+
+let count = 0 // total number of tests.
```
Comment (patak-dev) [inline docs/api/index.md:174]: We should remove the dots in all. Maybe the beforeEach example could be reworked to show something different.

**Comment by patak-dev** [comment]
@Snugug should we add examples for each option in the API section? Maybe it is better to move these to a guide and keep the API section compact?

File: docs/api/index.md (hunk 1 above)
Comment (dammy001) [reply docs/api/index.md:174]: @patak-dev okay. I'll update that now

Comment (dammy001) [reply docs/api/index.md]: @patak-dev checkout the new beforeEach example

**Comment by Snugug** [comment]
> @Snugug should we add examples for each option in the API section? Maybe it is better to move these to a guide and keep the API section compact?

I think having quick examples in the API section is a good thing; I personally find API documentation that's just types and arguments to be super unhelpful. As such, I think adding examples here are a net positive for the docs. To your point @patak-dev, I think it would be good to have a guide for these items, explaining them more in-depth, and have these examples be quick. 

That said, I think these examples could be improved; the way they're written, if I skim them quickly, it looks like I can actually get something out of these because they're returning functions. I think to make it clear there's no way to get something in or out of these functions, they shouldn't return anything, but rather using `async/await` to wait for a promise to finish, or just call a function "with side effects" but without returning anything.

As an aside, while this is being updated, can we remove "global" from the first sentence? We're not showing them as globals.

**Comment by dammy001** [comment]
@patak-dev setup and teardown APIs have been updated. Check it out

File: docs/api/index.md [hunk 2]
```diff
@@ -159,30 +159,96 @@ When you use `test` in the top level of file, they are collected as part of the
...
 ### beforeEach
 
-- **Type:** `beforeEach(fn: () => Awaitable<void>, timeout: number)`
+- **Type:** `beforeEach(fn: () => Awaitable<void>, timeout?: number)`
 
-  Register a callback to be called before each test in the current context.
+  Register a callback to be called before each of the tests in the current context runs.
+  If the function returns a promise, Vitest waits until the promise resolve before running the test.
+
+  Optionally, you can a timeout (in milliseconds) for specifying how long to wait before terminating. The default is 5 seconds.
```
Comment (patak-dev) [inline docs/api/index.md:171]:
```suggestion
  Optionally, you can pass a timeout (in milliseconds) defining how long to wait before terminating. The default is 5 seconds.
```

File: docs/api/index.md [hunk 3]
```diff
@@ -159,30 +159,96 @@ When you use `test` in the top level of file, they are collected as part of the
...
+
+  beforeEach(() => {
+    // Clear mocks and add some testing data after before each test run
+    stopMocking().then(() => {
+      addUser({ name: 'Patak '})
+    })
+  })
+
+  // Hook accepting a timeout
+  beforeEach(() => { }, 1000)
```
Comment (patak-dev) [inline docs/api/index.md:184]: I think we should only include an `async` hook directly, and avoid the second with the timeout (it should be clear from the signature already). And let's use a generic name for the user. Same for other examples
```suggestion
  import { beforeEach } from 'vitest'

  beforeEach(async () => {
    // Clear mocks and add some testing data after before each test run
    await stopMocking()
    addUser({ name: 'John'})
  })
```

**Review by patak-dev — CHANGES_REQUESTED** [review]

File: docs/api/index.md (hunk 3 above)
Comment (dammy001) [reply docs/api/index.md]: yeah. This makes more sense

**Review by sheremet-va — APPROVED** [review]

File: docs/api/index.md [hunk 4]
```diff
@@ -159,30 +159,92 @@ When you use `test` in the top level of file, they are collected as part of the
...
+
+  ```ts
+  import { afterEach } from 'vitest'
+
+  afterEach(() => {
+    clearTestingData() // clear testing data after each test run
+  })
+
+  // Hook accepting a timeout
+  afterEach(() => { }, 1000)
```
Comment (patak-dev) [inline docs/api/index.md:202]:
```suggestion
  afterEach(async () => {
    await clearTestingData() // clear testing data after each test run
  })
```

File: docs/api/index.md [hunk 5]
```diff
@@ -159,30 +159,89 @@ When you use `test` in the top level of file, they are collected as part of the
...
+
+  ```ts
+  import { beforeAll } from 'vitest'
+
+  beforeAll(() => {
+    startMocking() // called once before all tests run
+  })
+
+  // Hook accepting a timeout
+  beforeAll(() => { }, 1000)
```
Comment (patak-dev) [inline docs/api/index.md:220]:
```suggestion
  beforeAll(async () => {
    await startMocking() // called once before all tests run
  })
```

File: docs/api/index.md [hunk 6]
```diff
@@ -159,30 +159,89 @@ When you use `test` in the top level of file, they are collected as part of the
...
+
+  ```ts
+  import { afterAll } from 'vitest'
+
+  afterAll(() => {
+    stopMocking() // this method is called after all tests run
+  })
+
+  // Hook accepting a timeout
+  afterAll(() => { }, 1000)
```
Comment (patak-dev) [inline docs/api/index.md:239]:
```suggestion
  afterAll(async () => {
    await stopMocking() // this method is called after all tests run
  })
```

**Review by patak-dev — APPROVED** [review]

**Pull request merged**